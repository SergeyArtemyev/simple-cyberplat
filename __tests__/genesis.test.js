const axios = require('axios')
const fsp = require('fs').promises
const { Genesis, requestFromObj, responseToObj } = require('../index')
jest.mock('axios')

// init genesis client
async function getGenesis () {
  return new Genesis({
    defaultFields: {
      AP: 1111
    },
    provider: {
      url: 'https://predproc.tothemars.io/pay'
    },
    crypto: {
      key: await fsp.readFile('../simple-cyberplat/secret/key.pem', 'utf-8'),
      passphrase: ''
    }
  })
}

describe('genesis.request', () => {
  afterEach(() => {
    jest.restoreAllMocks()
  })

  beforeEach(() => {
    // config for mocking axios
    const response = { RESULT: '1', ERROR: '3' }
    // mock axios return value
    axios.request.mockResolvedValue(response)
  })
  // need to think about proper names for tests
  test('axios should be called with certain data', async () => {
    // config for mocking axios
    const answer = { RESULT: '1', ERROR: '3' }
    // config for data which axios should be called with
    const data = requestFromObj({ AP: 1111, REQ_TYPE: 12 })
    // get genesis client
    const transport = await getGenesis()
    // actual call
    const result = await transport.request({
      REQ_TYPE: 12
    })
    // use expect.objectContaining to check if certain data exists in request object
    expect(axios.request).toHaveBeenCalledWith(expect.objectContaining({ data }))
    expect(axios.request).toHaveBeenCalledTimes(1)
    expect(result).toEqual(answer)
  })

  test('axios return value should be decoded', async () => {
    const answer = { RESULT: '1', ERROR: '3' }
    const encodedAnswer = requestFromObj(answer)
    const decodedAnswer = responseToObj(encodedAnswer)
    // prepare variable for a func
    let transformResponse
    // create spy
    jest.spyOn(axios, 'request').mockImplementationOnce((options) => {
      // assign transformResponse function from the actual call config
      transformResponse = options.transformResponse
      // why return promise?(i get it from stackoverflow example) it works fine without it
      // return Promise.resolve({})
    })

    // get genesis client
    const transport = await getGenesis()
    // actual call
    await transport.request({
      REQ_TYPE: 12
    })
    // maybe this is an unnecessary call
    expect(axios.request).toBeCalled()
    // get decoded response with the transformResponse function from the config of the axios.request call
    const response = transformResponse(encodedAnswer)
    // check response
    expect(response).toEqual(decodedAnswer)
    // it is called two times because there are two tests
    // i need to find a way to isolate calls
    expect(axios.request).toBeCalledTimes(2)
  })
})
