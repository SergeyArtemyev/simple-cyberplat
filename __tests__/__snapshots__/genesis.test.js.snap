// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utf8-win1251 transcoding 1`] = `
Object {
  "data": Object {
    "data": Array [
      65,
      80,
      61,
      49,
      49,
      49,
      49,
      13,
      10,
      82,
      69,
      81,
      95,
      84,
      89,
      80,
      69,
      61,
      236,
      224,
      236,
      224,
      32,
      236,
      251,
      235,
      224,
      32,
      240,
      224,
      236,
      243,
    ],
    "type": "Buffer",
  },
  "headers": Object {
    "Content-Type": "text/plain; encoding=windows-1251",
    "x-sign": "CksfMGXRE51QzUnvEeTzFCcDm7lsevqgc3drhcPQQQpDH2PndfUKiq11lJhao5ai%2FzIAjFINxGLDR2yxSAnyl4KunFLuCGzPMBTwkYqrTjGFOPDyAuy6a9vwQ1OA9qGdc7gZPAVe65TzsMjJ1CXyXkvWphjpqCehLYpqnoyXRGs%3D",
  },
  "method": "POST",
  "responseType": "arraybuffer",
  "transformResponse": [Function],
  "url": "https://predproc.tothemars.io/pay",
}
`;

exports[`utf8-win1251 transcoding 2`] = `
Object {
  "BAR": "зайчики заплакали",
  "FOO": "уточки закрякали",
}
`;
